README
This is release 0.2 of lizip.
See doc/manual.html for documentation and licenses.


NEWS
2009-06-25  Release 0.2
  * The latest Info-ZIP DLLs (zip32z64 v3.0 and unzip32 v6.0) are supported
    and required.

2009-06-18  Release 0.1  
  * First public release.
  * The following DLLs are supported: zip32 v2.32 and unzip32 v5.52.



src/lzip.c
/*
  started    : 2009-06-02
  written by : Shmuel Zeigerman
*/

#include <zip/windll/structs.h>
#include "common.h"


extern int luaopen_lunzip (lua_State *);


/* Must use a global, since Info-ZIP callbacks don't provide
 * for an object pointer
 * **********************************************************/
static lua_State *GL;


/* Password entry routine - see password.c in the wiz directory for how
   this is actually implemented in Wiz.
 */
static int WINAPI CallbackPassword
    (LPSTR pwbuf, int bufsize, LPCSTR mode, LPCSTR name)
{
  return callback_password(GL, pwbuf, bufsize, mode, name);
}


/* Callback "print" routine */
static int WINAPI CallbackPrint(char *buf, unsigned long size)
{
  return callback_print(GL, buf, size);
}


/* Callback "comment" routine. See comment.c in the wiz directory for how
   this is actually implemented in Wiz.
 */
static int WINAPI CallbackComment (char *szBuf)
{
  szBuf[0] = '\0';
  lua_getfield(GL, LUA_ENVIRONINDEX, "comment");
  if (lua_isfunction(GL, -1)) {
    lua_pushstring(GL, szBuf);
    if (0 == lua_pcall(GL, 1, 1, 0)) {
      if (lua_isstring(GL, -1)) {
        strncpy(szBuf, lua_tostring(GL,-1), 0xFFF0);
        szBuf[0xFFF0] = '\0';
      }
    }
  }
  lua_pop(GL,1);
  return TRUE;
}


static int WINAPI CallbackServiceApplication
    (LPCSTR name, unsigned __int64 size)
{
  int ret = 0;
  lua_getfield(GL, LUA_ENVIRONINDEX, "ServiceApplication");
  if (lua_isfunction(GL, -1)) {
    lua_pushstring(GL, name);
    lua_pushnumber(GL, size);
    if (0 == lua_pcall(GL, 2, 1, 0))
      ret = lua_toboolean(GL,-1);
  }
  lua_pop(GL,1);
  return ret;
}


static ZIPUSERFUNCTIONS ZipUserFunctions = {
  CallbackPrint, CallbackComment, CallbackPassword, NULL,
  CallbackServiceApplication
};


static int f_version (lua_State *L) {
  ZpVer ver;

  GL = L;
  memset(&ver, 0, sizeof(ZpVer));
  ver.structlen = sizeof(ZpVer);
  ZpVersion(&ver);

  lua_createtable(L, 0, 9);
  /* ---------------------------------------------------------------- */
  set_boolean_field(L, "is_beta", ver.flag & 0x01);
  set_boolean_field(L, "uses_zlib", ver.flag & 0x02);
  set_string_field(L, "betalevel", ver.betalevel);
  set_string_field(L, "date", ver.date);
  set_string_field(L, "zlib_version", ver.zlib_version);
  set_boolean_field(L, "fEncryption", ver.fEncryption);
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver.zip.major, (int)ver.zip.minor,
                                 (int)ver.zip.patchlevel);
  lua_setfield(L, -2, "zip");
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver.os2dll.major, (int)ver.os2dll.minor,
                                 (int)ver.os2dll.patchlevel);
  lua_setfield(L, -2, "os2dll");
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver.windll.major, (int)ver.windll.minor,
                                 (int)ver.windll.patchlevel);
  lua_setfield(L, -2, "windll");
  /* ---------------------------------------------------------------- */
  return 1;
}


static int f_init (lua_State *L) {
  if (lua_isnoneornil(L, 1)) {
    lua_createtable(L, 0, 0);
    lua_insert(L, 1);
  }
  luaL_checktype(L, 1, LUA_TTABLE);
  optfunction(L, 1, "print");
  optfunction(L, 1, "comment");
  optfunction(L, 1, "password");
  optfunction(L, 1, "ServiceApplication");
  lua_pushboolean(L, ZpInit(&ZipUserFunctions));
  return 1;
}


static void setoptions (lua_State *L, int pos, ZPOPT *opt) {
  int fLevel;
  luaL_checktype(L, pos, LUA_TTABLE);

  optstring(L, pos, "Date", (const char**)&opt->Date);
  optstring(L, pos, "szRootDir", (const char**)&opt->szRootDir);
  optstring(L, pos, "szTempDir", (const char**)&opt->szTempDir);
  optboolean(L, pos, "fTemp", &opt->fTemp);
  optboolean(L, pos, "fSuffix", &opt->fSuffix);
  optboolean(L, pos, "fEncrypt", &opt->fEncrypt);
  optboolean(L, pos, "fSystem", &opt->fSystem);
  optboolean(L, pos, "fVolume", &opt->fVolume);
  optboolean(L, pos, "fExtra", &opt->fExtra);
  optboolean(L, pos, "fNoDirEntries", &opt->fNoDirEntries);
  optboolean(L, pos, "fExcludeDate", &opt->fExcludeDate);
  optboolean(L, pos, "fIncludeDate", &opt->fIncludeDate);
  optboolean(L, pos, "fVerbose", &opt->fVerbose);
  optboolean(L, pos, "fQuiet", &opt->fQuiet);
  optboolean(L, pos, "fCRLF_LF", &opt->fCRLF_LF);
  optboolean(L, pos, "fLF_CRLF", &opt->fLF_CRLF);
  optboolean(L, pos, "fJunkDir", &opt->fJunkDir);
  optboolean(L, pos, "fGrow", &opt->fGrow);
  optboolean(L, pos, "fForce", &opt->fForce);
  optboolean(L, pos, "fMove", &opt->fMove);
#if 0  /* actions */
  optboolean(L, pos, "fDeleteEntries", &opt->fDeleteEntries);
  optboolean(L, pos, "fUpdate", &opt->fUpdate);
  optboolean(L, pos, "fFreshen", &opt->fFreshen);
#endif
  optboolean(L, pos, "fJunkSFX", &opt->fJunkSFX);
  optboolean(L, pos, "fLatestTime", &opt->fLatestTime);
  optboolean(L, pos, "fComment", &opt->fComment);
  optboolean(L, pos, "fOffsets", &opt->fOffsets);
  optboolean(L, pos, "fPrivilege", &opt->fPrivilege);
  /* optboolean(L, pos, "fEncryption", &opt->fEncryption); */ /* read-only flag */
  optstring (L, pos, "szSplitSize", (const char**)&opt->szSplitSize);
  optinteger(L, pos, "fRecurse", &opt->fRecurse);
  optinteger(L, pos, "fRepair", &opt->fRepair);

  fLevel = -1;
  optinteger(L, pos, "fLevel", &fLevel);
  if (fLevel >= 0 && fLevel <= 9)
    opt->fLevel = fLevel + '0';
}


static int archive (lua_State *L, int action) {
  ZCL zcl;
  ZPOPT opt;
  int result;
  const char *argv[2];

  GL = L;
  lua_settop(L, 3);

  /* 1-st argument */
  memset(&zcl, 0, sizeof(ZCL));
  zcl.lpszZipFN = (char*) luaL_checkstring(L, 1);

  /* 2-nd argument */
  if (lua_isstring(L, 2)) {
    argv[0] = lua_tostring(L, 2);
    argv[1] = 0;
    zcl.argc = 1;
    zcl.FNV = (char**)argv;
  }
  else if (lua_istable(L, 2))
    create_argarray(L, 2, &zcl.argc, &zcl.FNV);
  else
    return luaL_typerror(L, 2, "string or table");

  /* 3-rd argument */
  memset(&opt, 0, sizeof(ZPOPT));
  opt.fLevel = '6';
  if (!lua_isnil(L, 3))
    setoptions(L, 3, &opt);

  opt.fDeleteEntries = opt.fUpdate = opt.fFreshen = 0;
  if (action == 'u') opt.fUpdate = 1;
  else if (action == 'f') opt.fFreshen = 1;
  else if (action == 'd') opt.fDeleteEntries = 1;

  result = ZpArchive(zcl, &opt);

  lua_pushinteger(L, result);
  if (result >= 0 && result <= ZE_MAXERR)
    lua_pushstring(L, ziperrors[result].string);
  else
    lua_pushstring(L, "unknown error code");
  return 2;
}


static int f_add     (lua_State *L)  { return archive(L, 'a'); }
static int f_update  (lua_State *L)  { return archive(L, 'u'); }
static int f_freshen (lua_State *L)  { return archive(L, 'f'); }
static int f_delete  (lua_State *L)  { return archive(L, 'd'); }


static const luaL_reg lib[] = {
  {"init",         f_init},
  {"version",      f_version},

  {"add",          f_add},
  {"update",       f_update},
  {"freshen",      f_freshen},
  {"delete",       f_delete},

  {"filetime",     f_filetime},

  {NULL,           NULL}
};


int luaopen_lizip (lua_State *L)
{
  ZpInit(&ZipUserFunctions); /* prevents crash */

  lua_newtable(L);
  lua_replace(L, LUA_ENVIRONINDEX); /* used in callbacks */
  luaL_register(L, "zip", lib);

  luaopen_lunzip(L);

  return 0;
}


src/common.c
#include "common.h"
#include <zip/windll/structs.h>


static void checkfornil (lua_State *L, const char *key) {
  if (!lua_isnil(L,-1))
    luaL_argerror(L, 1, lua_pushfstring(L, "invalid type of field '%s'", key));
}


void optstring (lua_State *L, int pos, const char* key, const char **trg) {
  lua_getfield(L, pos, key);
  if (lua_isstring(L,-1))
    *trg = lua_tostring(L, -1);
  else
    checkfornil(L, key);
  lua_pop(L,1);
}


void optboolean (lua_State *L, int pos, const char* key, int *trg) {
  lua_getfield(L, pos, key);
  if (lua_type(L,-1) == LUA_TBOOLEAN)
    *trg = lua_toboolean(L, -1);
  else
    checkfornil(L, key);
  lua_pop(L,1);
}


void optinteger (lua_State *L, int pos, const char* key, int *trg) {
  lua_getfield(L, pos, key);
  if (lua_isnumber(L,-1))
    *trg = lua_tointeger(L, -1);
  else
    checkfornil(L, key);
  lua_pop(L,1);
}


void optfunction (lua_State *L, int pos, const char* key) {
  lua_getfield(L, pos, key);
  if (!lua_isfunction(L,-1))
    checkfornil(L, key);
  lua_setfield(L, LUA_ENVIRONINDEX, key);
}


/*
 * Creates an argarray from a Lua table.
 * Replaces '\\' to '/' in all items (unzip32.dll wouldn't find items with '\\')
 * Side effect: pushes a userdata and a table onto Lua stack (but don't remove
 * them manually from there!)
 */
void create_argarray (lua_State *L, int pos, int *argc, char ***argv) {
  int i;
  if (pos < 0)
    pos += lua_gettop(L) + 1;
  luaL_checktype(L, pos, LUA_TTABLE);
  *argc = lua_objlen(L, pos);
  *argv = (char**) lua_newuserdata(L, (*argc+1) * sizeof(char*));
  lua_createtable(L, *argc, 0);
  for (i=0; i<*argc; i++, lua_pop(L,1)) {
    const char *item;
    lua_pushinteger(L, i+1);
    lua_gettable(L, pos);
    item = lua_tostring(L, -1);
    if (item == NULL)
      luaL_error(L, "arg #d: invalid type on position %d", pos, i+1);
    item = luaL_gsub(L, item, "\\", "/");
    (*argv)[i] = (char*) item;
    lua_rawseti(L, -3, i+1);
  }
  (*argv)[*argc] = NULL;
}


void optional_argarray (lua_State *L, int pos, int *argc, char ***argv) {
  if (lua_istable(L, pos))
    create_argarray(L, pos, argc, argv);
  else if (!lua_isnoneornil(L, pos))
    luaL_typerror(L, pos, "table or nil");
}


void set_string_field (lua_State *L, const char* key, const char* val) {
  lua_pushstring(L, val);
  lua_setfield(L, -2, key);
}


void set_boolean_field (lua_State *L, const char* key, int val) {
  lua_pushboolean(L, val);
  lua_setfield(L, -2, key);
}


void set_integer_field (lua_State *L, const char* key, int val) {
  lua_pushinteger(L, val);
  lua_setfield(L, -2, key);
}


void set_number_field (lua_State *L, const char* key, double val) {
  lua_pushnumber(L, val);
  lua_setfield(L, -2, key);
}


int callback_password (lua_State *L, LPSTR pwbuf, int bufsize, LPCSTR param,
  LPCSTR filename)
{
  int status = IZ_PW_ERROR;
  lua_getfield(L, LUA_ENVIRONINDEX, "password");
  if (!lua_isfunction(L,-1))
    status = IZ_PW_CANCELALL;
  else {
    lua_pushinteger(L, bufsize);
    lua_pushstring(L, param);
    lua_pushstring(L, filename);
    if (0 == lua_pcall(L, 3, 1, 0)) {
      if (lua_isstring(L,-1)) {
        const char *str = lua_tostring(L,-1);
        int objlen = lua_objlen(L,-1);
        if (objlen < bufsize && objlen == strlen(str)) {
          strcpy(pwbuf, str);
          status = IZ_PW_ENTERED;
        }
      }
      else if (lua_isnil(L,-1))
        status = IZ_PW_CANCEL;
    }
  }
  lua_pop(L,1);
  return status;
}


int callback_print (lua_State *L, char *buf, unsigned long size)
{
  int ret = (unsigned int) size;
  lua_getfield(L, LUA_ENVIRONINDEX, "print");
  if (lua_isfunction(L, -1)) {
    lua_pushlstring(L, buf, size);
    if (0 == lua_pcall(L, 1, 1, 0)) {
      if (lua_isnumber(L, -1)) {
        ret = lua_tointeger(L, -1);
      }
    }
  }
  lua_pop(L,1);
  return ret;
}

/* Get or set time of file's last modification. Time unit is second.
 * (Windows-specific)
 */
static int FileTime (const char *filename, double *time, int get)
{
  FILETIME t_write;
  LARGE_INTEGER LI;
  HANDLE handle;

  handle = CreateFile(filename, get ? GENERIC_READ:GENERIC_WRITE, 0, NULL,
    OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
  if (handle == INVALID_HANDLE_VALUE)
    return 0;

  if (get) {
    if (!GetFileTime(handle, NULL, NULL, &t_write)) {
      CloseHandle(handle);
      return 0;
    }
    memcpy(&LI, &t_write, sizeof(LARGE_INTEGER));
    *time = LI.QuadPart / 10000000;       /* convert 100 ns units to seconds */
    CloseHandle(handle);
    return 1;
  }

  LI.QuadPart = *time * 10000000;         /* convert seconds to 100 ns units */
  memcpy(&t_write, &LI, sizeof(FILETIME));
  if (!SetFileTime(handle, NULL, NULL, &t_write)) {
    CloseHandle(handle);
    return 0;
  }
  CloseHandle(handle);
  return 1;
}

int f_filetime (lua_State *L) {
  const char *filename = luaL_checkstring(L, 1);
  double time;
  if (lua_isnoneornil(L, 2)) {
    if (FileTime(filename, &time, 1))
      lua_pushnumber(L, time);
    else
      lua_pushnil(L);
    return 1;
  }
  time = luaL_checknumber(L, 2);
  lua_pushboolean (L, FileTime(filename, &time, 0));
  return 1;
}



src/common.h
#ifndef LIZIP_COMMON_H
#define LIZIP_COMMON_H

#include <lua.h>
#include <lauxlib.h>

void optinteger  (lua_State *L, int pos, const char* key, int *trg);
void optstring   (lua_State *L, int pos, const char* key, const char **trg);
void optboolean  (lua_State *L, int pos, const char* key, int *trg);
void optfunction (lua_State *L, int pos, const char* key);
void create_argarray   (lua_State *L, int pos, int *argc, char ***argv);
void optional_argarray (lua_State *L, int pos, int *argc, char ***argv);
void set_string_field  (lua_State *L, const char* key, const char* val);
void set_boolean_field (lua_State *L, const char* key, int val);
void set_integer_field (lua_State *L, const char* key, int val);
void set_number_field  (lua_State *L, const char* key, double val);
int  callback_password (lua_State *L, char* pwbuf, int bufsize,
                        const char* param, const char* filename);
int  callback_print    (lua_State *L, char *buf, unsigned long size);
int  f_filetime        (lua_State *L);

#endif



src/lunzip.c
/*
  started    : 2009-06-03
  written by : Shmuel Zeigerman
*/

#include <windows.h>
#include <unzip/windll/structs.h>
#include <unzip/windll/decs.h>
#include "common.h"


/* Must use a global, since Info-ZIP callbacks don't provide
 * for an object pointer
 * **********************************************************/
static lua_State *GL;


/* Password entry routine - see password.c in the wiz directory for how
   this is actually implemented in WiZ.
 */
static int WINAPI CallbackPassword (LPSTR pwbuf, int bufsize, LPCSTR prompt,
    LPCSTR filename)
{
  return callback_password(GL, pwbuf, bufsize, prompt, filename);
}


static int WINAPI CallbackPrint (LPSTR buf, unsigned long size)
{
  return callback_print(GL, buf, size);
}


static void WINAPI CallbackSound ()
{
  lua_getfield(GL, LUA_ENVIRONINDEX, "sound");
  if (lua_isfunction(GL, -1)) {
    if (0 == lua_pcall(GL, 0, 0, 0))
      ;
  }
  lua_pop(GL,1);
}


static int WINAPI CallbackReplace (LPSTR efnam, unsigned efbufsiz) {
  /* This is where you will decide if you want to replace, rename etc existing
     files.
   */
  int ret = IDM_REPLACE_NO;
  lua_getfield(GL, LUA_ENVIRONINDEX, "replace");
  if (lua_isfunction(GL, -1)) {
    lua_pushstring(GL, efnam);
    lua_pushinteger(GL, efbufsiz);
    if (0 == lua_pcall(GL, 2, 2, 0)) {
      const char* str = lua_tostring(GL, -2);
      if (str) {
        if (!strcmp(str, "no")) ret = IDM_REPLACE_NO;
        else if (!strcmp(str, "yes")) ret = IDM_REPLACE_YES;
        else if (!strcmp(str, "all")) ret = IDM_REPLACE_ALL;
        else if (!strcmp(str, "none")) ret = IDM_REPLACE_NONE;
        else if (!strcmp(str, "rename") && lua_isstring(GL, -1)) {
          ret = IDM_REPLACE_RENAME;
          strncpy(efnam, lua_tostring(GL, -1), efbufsiz - 1);
          efnam[efbufsiz - 1] = '\0';
        }
      }
      lua_pop(GL,1);
    }
  }
  lua_pop(GL,1);
  return ret;
}


/* Essentially what this function is for is to do a listing of an archive
   contents.
 */
static void WINAPI CallbackSendApplicationMessage (
    z_uint8 ucsize,
    z_uint8 csiz,
    unsigned cfactor,
    unsigned mo, unsigned dy, unsigned yr, unsigned hh, unsigned mm,
    char c, LPCSTR filename, LPCSTR methbuf, unsigned long crc, char fCrypt)
{
  lua_getfield(GL, LUA_ENVIRONINDEX, "SendApplicationMessage");
  if (lua_isfunction(GL, -1)) {
    lua_pushnumber(GL, ucsize);
    lua_pushnumber(GL, csiz);
    lua_pushnumber(GL, cfactor);
    lua_pushnumber(GL, mo);
    lua_pushnumber(GL, dy);
    lua_pushnumber(GL, yr);
    lua_pushnumber(GL, hh);
    lua_pushnumber(GL, mm);
    lua_pushlstring(GL, &c, 1);
    lua_pushstring(GL, filename);
    lua_pushstring(GL, methbuf);
    lua_pushnumber(GL, crc);
    lua_pushlstring(GL, &fCrypt, 1);
    if (0 == lua_pcall(GL, 13, 0, 0))
      ;
  }
  lua_pop(GL,1);
}


static int WINAPI CallbackServCallBk (const char* filename, z_uint8 size)
{
  int ret = 0;
  lua_getfield(GL, LUA_ENVIRONINDEX, "ServCallBk");
  if (lua_isfunction(GL, -1)) {
    lua_pushstring(GL, filename);
    lua_pushnumber(GL, size);
    if (0 == lua_pcall(GL, 2, 1, 0))
      ret = lua_toboolean(GL, -1);
  }
  lua_pop(GL,1);
  return ret;
}


static USERFUNCTIONS get_user_functions () {
  USERFUNCTIONS uf;
  memset(&uf, 0, sizeof(USERFUNCTIONS));
  uf.password = CallbackPassword;
  uf.print = CallbackPrint;
  uf.sound = CallbackSound;
  uf.replace = CallbackReplace;
  uf.SendApplicationMessage = CallbackSendApplicationMessage;
  uf.ServCallBk = CallbackServCallBk;
  return uf;
}


static int f_noprinting (lua_State *L) {
  luaL_checkany(L, 1);
  Wiz_NoPrinting(lua_toboolean(L, 1));
  return 0;
}


static int f_validate (lua_State *L) {
  const char *archive = luaL_checkstring(L, 1);
  lua_pushinteger(L, Wiz_Validate((char*)archive, TRUE));
  return 1;
}


static int f_version (lua_State *L) {
  const UzpVer *ver = UzpVersion();

  lua_createtable(L, 0, 8);
  /* ---------------------------------------------------------------- */
  set_boolean_field(L, "is_beta", ver->flag & 0x01);
  set_boolean_field(L, "uses_zlib", ver->flag & 0x02);
  set_string_field(L, "betalevel", ver->betalevel);
  set_string_field(L, "date", ver->date);
  set_string_field(L, "zlib_version", ver->zlib_version);
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver->unzip.major, (int)ver->unzip.minor,
                                 (int)ver->unzip.patchlevel);
  lua_setfield(L, -2, "unzip");
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver->os2dll.major, (int)ver->os2dll.minor,
                                 (int)ver->os2dll.patchlevel);
  lua_setfield(L, -2, "os2dll");
  /* ---------------------------------------------------------------- */
  lua_pushfstring(L, "%d.%d.%d", (int)ver->windll.major, (int)ver->windll.minor,
                                 (int)ver->windll.patchlevel);
  lua_setfield(L, -2, "windll");
  /* ---------------------------------------------------------------- */
  return 1;
}


static void get_options_from_table (lua_State *L, int pos, DCL *dcl)
{
  optboolean(L, pos, "ExtractOnlyNewer", &dcl->ExtractOnlyNewer);
  optboolean(L, pos, "SpaceToUnderscore", &dcl->SpaceToUnderscore);
  optboolean(L, pos, "PromptToOverwrite", &dcl->PromptToOverwrite);

  optinteger(L, pos, "fQuiet", &dcl->fQuiet);

  optboolean(L, pos, "ncflag", &dcl->ncflag);
  optboolean(L, pos, "ntflag", &dcl->ntflag);
  optboolean(L, pos, "nvflag", &dcl->nvflag);
  optboolean(L, pos, "nfflag", &dcl->nfflag);
  optboolean(L, pos, "nzflag", &dcl->nzflag);

  optinteger(L, pos, "ndflag", &dcl->ndflag);

  optboolean(L, pos, "noflag", &dcl->noflag);
  optboolean(L, pos, "naflag", &dcl->naflag);
  optboolean(L, pos, "nZIflag", &dcl->nZIflag);
  optboolean(L, pos, "B_flag", &dcl->B_flag);
  optboolean(L, pos, "C_flag", &dcl->C_flag);

  optinteger(L, pos, "D_flag", &dcl->D_flag);
  optinteger(L, pos, "U_flag", &dcl->U_flag);
  optinteger(L, pos, "fPrivilege", &dcl->fPrivilege);

  /* optstring(L, pos, "lpszZipFN", (const char **) &dcl->lpszZipFN); */
  /* optstring(L, pos, "lpszExtractDir", (const char **) &dcl->lpszExtractDir); */
}


static void get_options_from_string (lua_State *L, int pos, DCL *dcl)
{
  const char *s = lua_tostring(L, pos);

  /* process 'freshen' and 'update' */
  if (strchr(s, 'f'))
    dcl->nfflag = 1;
  if (strchr(s, 'u')) {
    if (dcl->nfflag)
      luaL_argerror(L, pos, "'f' and 'u' options are mutually exclusive");
    dcl->ExtractOnlyNewer = 1;
  }

  /* process 'overwrite' and 'never overwrite' */
  if (strchr(s, 'o'))
    dcl->noflag = 1;
  if (strchr(s, 'n')) {
    if (dcl->noflag)
      luaL_argerror(L, pos, "'n' and 'o' options are mutually exclusive");
  }
  else if (!dcl->noflag)
    dcl->PromptToOverwrite = 1;
}


static int unzip_common (lua_State *L, int action) {
  const char *ZipFN;
  int ifnc = 0, xfnc = 0;
  char **ifnv = 0, **xfnv = 0;
  DCL dcl;
  int pos = 1;
  int result;
  USERFUNCTIONS uf = get_user_functions();

  GL = L;
  lua_settop(L, 5);

  ZipFN = luaL_checkstring(L, pos++);

  memset(&dcl, 0, sizeof(DCL));
  if (action == 'u') {
    const char *ExtractDir = luaL_optstring(L, pos++, NULL);

    if (lua_istable(L, pos)) {
      get_options_from_table(L, pos, &dcl);
      dcl.nvflag = dcl.ntflag = 0;
    }
    else if (lua_isstring(L, pos))
      get_options_from_string(L, pos, &dcl);
    else if (!lua_isnil(L, pos))
      return luaL_typerror(L, pos, "table or nil");

    dcl.lpszExtractDir = (char*) ExtractDir;
    pos++;
  }
  else if (action == 'l')
    dcl.nvflag = 1;
  else if (action == 't')
    dcl.ntflag = 1;

  dcl.StructVersID = UZ_DCL_STRUCTVER;
  dcl.lpszZipFN = (char*) ZipFN;

  optional_argarray(L, pos++, &ifnc, &ifnv);
  optional_argarray(L, pos++, &xfnc, &xfnv);

  result = Wiz_SingleEntryUnzip(ifnc, ifnv, xfnc, xfnv, &dcl, &uf);
  lua_pushinteger(L, result);

  if (result == PK_OK && dcl.nvflag) {
    lua_createtable(L, 0, 5);
    set_number_field (L, "TotalSizeComp", uf.TotalSizeComp);
    set_number_field (L, "TotalSize",     uf.TotalSize);
    set_number_field (L, "CompFactor",    uf.CompFactor);
    set_number_field (L, "NumMembers",    uf.NumMembers);
    set_number_field (L, "cchComment",    uf.cchComment);
    return 2;
  }
  return 1;
}


static int f_unzip (lua_State *L)   { return unzip_common(L, 'u'); }
static int f_list  (lua_State *L)   { return unzip_common(L, 'l'); }
static int f_test  (lua_State *L)   { return unzip_common(L, 't'); }


static int f_unziptomemory (lua_State *L) {
  UzpBuffer buff;
  const char *zip, *file;
  int ret;
  USERFUNCTIONS uf = get_user_functions();

  GL = L;
  zip = luaL_checkstring(L, 1);
  file = luaL_checkstring(L, 2);
  file = luaL_gsub(L, file, "\\", "/");   /* workaround: bug in unzip32.dll */
  ret = Wiz_UnzipToMemory((char*)zip, (char*)file, &uf, &buff);
  if (ret == 1) {
    lua_pushlstring(L, buff.strptr, buff.strlength);
    UzpFreeMemBuffer(&buff);
  }
  else
    lua_pushnil(L);
  return 1;
}


static int f_init (lua_State *L) {
  if (lua_isnoneornil(L, 1)) {
    lua_createtable(L, 0, 0);
    lua_insert(L, 1);
  }
  luaL_checktype(L, 1, LUA_TTABLE);
  optfunction(L, 1, "print");
  optfunction(L, 1, "sound");
  optfunction(L, 1, "replace");
  optfunction(L, 1, "password");
  optfunction(L, 1, "SendApplicationMessage");
  optfunction(L, 1, "ServCallBk");
  return 0;
}


static const luaL_reg lib[] = {
  {"version",       f_version},
  {"init",          f_init},
  {"unzip",         f_unzip},
  {"list",          f_list},
  {"test",          f_test},
  {"unziptomemory", f_unziptomemory},
  {"validate",      f_validate},
  {"noprinting",    f_noprinting},
  {"filetime",      f_filetime},
  {NULL,           NULL}
};


int luaopen_lunzip (lua_State *L) {
  lua_newtable(L);
  lua_replace(L, LUA_ENVIRONINDEX); /* used in callbacks */
  luaL_register(L, "unzip", lib);
  return 0;
}
